File: 7KYU/Simple directions reversal/solution.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    private static string Swap(string dir)
    => dir == "Right" ? "Left" : "Right";

    public static string [] solve(string [] arr)
    {
      var directions = arr.Select(a => a.Split(" on "));
      var turns = new List<string>{"Begin"};
      turns.AddRange(directions.Select(x => x[0])
                               .Skip(1)
                               .Select(Swap)
                               .Reverse()
                               .ToList());
      var roads = directions.Select(x => x[1])
                            .Reverse();
      return roads.Zip(turns, (r, t) => t + " on " + r).ToArray();
    }
}

File: 7KYU/Simple directions reversal/solution.clj
(ns kata
  (:require [clojure.string :as str]))

(defn swap-directions [turn]
  (cond (= "Right" turn) "Left"
        (= "Left" turn) "Right"
        :else "Begin"))

(defn reverse-after [n xs]
  (concat (take n xs) (reverse (drop n xs))))

(defn solve [dirs]
  (let [directions (map (fn [d] (str/split d #" on ")) dirs)
        turns (->> (map first directions)
                   (map swap-directions)
                   ((partial reverse-after 1)))
        streets (reverse (map second directions))]
    (map (fn [t s] (str t " on " s)) turns streets)))